[proposal] Composite Features (#208)
devcontainers/spec: proposals/features-composite-dependencies.md

# Composite Features

Related: https://github.com/devcontainers/spec/issues/109

## Motivation

> We've seen significant interest in the ability to "reuse" or "extend" a given Feature with one or more additional Features.  Often in software a given tool depends on another (or several) tool(s)/framework(s).  As the dev container Feature ecosystem has grown, there has been a growing need to reduce redundant code in Features by utilizing first installing a set of dependant Features.

Reviewer: ```suggestion
We've seen significant interest in the ability to "reuse" or "extend" a given Feature with one or more additional Features.  Often in software a given tool depends on another (or several) tool(s)/framework(s).  As the dev container Feature ecosystem has grown, there has been a growing need to reduce redundant code in Features by first installing a set of dependent Features.
```

> 

Reviewer: Would we want to link to https://github.com/devcontainers/spec/pull/209 as well, since both are related to sharing or reusing Feature code in some way?

## Goal

> The solution shall provide a way to publish a Feature that "depends" on >= 1 other pubished Features. Dependent Features will be installed by the ochestrating tool, with the version and order set by the author if necessary.

Reviewer: ```suggestion
The solution shall provide a way to publish a Feature that "depends" on >= 1 other published Features. Dependent Features will be installed by the orchestrating tool, with the version and order set by the author if necessary.
```

### Direct curl

We've seen instances where users directly `curl` a Feature's `install.sh` script to `bash`.  This strategy also doesn't merge in the other dev container configuration properties that a Feature may declare.

### Additional inspiration

Inspiration was taken from [this spec issue on the topic](https://github.com/devcontainers/spec/issues/109), the repositories listed above, [the buildpack specification](https://docs.cloudfoundry.org/buildpacks/understand-buildpacks.html), and [VS Code extension packs](https://code.visualstudio.com/blogs/2017/03/07/extension-pack-roundup).

## Specification

Introduce a new file type `devcontainer-feature.composite.json` with the following properties.

| Property | Type | Description |
|----------|------|-------------|
| `id` | `string` | The ID of the Feature.  This follows the same semantics of the `id` property in the `devcontainer-feature.json` file. |
| `version` | `string` | The version of the Feature.  This follows the same semantics of the `version` property in the `devcontainer-feature.json` file. |

> | `features` | `array` | An array of objects (in installation order) that define the Feature(s) that compose this Feature. |

Reviewer 2: Does the correct installation order depend on each of the install script's requirements?

Author: We'd want to peek at each Feature's `installsAfter`, i think. That was the primary reason that I didn't want to (for now) enable nesting composite Features.   This spec does NOT let you nest one composite feature in another, making this process easier to resolve (we can "flatten" the installation from composite Features into one install order.

This is something to discuss and finalize in the spec, but we'd probably follow `installsAfter` installation order with precedence within a composite Feature first, and then some algorithm for ordering composite Features based on that?  Perhaps the other way around?

Reviewer 2: Since we are already thinking about composite features depending on each other: Wouldn't it be simpler to avoid the additional artifact type of a "composite feature" and instead allow features to have hard dependencies in addition to the soft dependencies supported with "installAfter"? We already compute the dependency graph including cycle detection for "installAfter". (Other package management systems also seem to have only one package type that can carry code/binaries and dependencies. Do we have an example that keeps these apart?)

We could add "dependencies" to the feature metadata as an object like "features" in the devcontainer.json, so installation order is decided by the same dependency graph algorithm. A dependency would implicitly go into the "installAfter" list of the feature depending on it.

> | `features.id` | `string` | The ID of the Feature that this Feature depends on. |

Reviewer 2: Can these be from a different repository/namespace?

Author: Yes, I see no reason to restrict if this proposal is followed.

> | `features.version` | `string` | The version of the Feature that this Feature depends on. |

Reviewer 2: What if a different version is configured in the devcontainer.json?

> | `features.detect` | `string` | A command that will be executed in a shell to determine if the Feature should be installed.  If the command returns a non-zero exit code, the Feature will be installed.  If the command returns a zero exit code, the remaining install steps will be skipped. |

Reviewer 2: Does "remaining install steps will be skipped" mean, the remaining features listed in the array will be skipped?

Author: I will clarify this - I meant that _each Feature_ will have a `detect` that will cause the further installation and merging of dev container config **for that Feature**.  Remaining Features in the array will still be executed (and their detect phase run, too).

Reviewer 2: What are the examples we want to cover with this? Would this run after all previous features have detected & installed, immediately before this feature's install script would run?

> | `features.options` | `object` | An object of key/value pairs that will be passed to the Feature's `install.sh` script. |

Reviewer 2: How are these merged when the same feature is installed by the devcontainer.json?

Author: Similar to https://github.com/devcontainers/spec/pull/208#discussion_r1140474890

#### Example `devcontainer-feature.composite.json`

```jsonc
{
    "id": "ghcr.io/devcontainers/features/composite",
    "version": "1.0.0",
    "features": [
        {
            "id": "ghcr.io/devcontainers/features/a", // Must be a standalone Feature. (A composite Feature cannot depend on a composite Feature).
            "version": "1.2.3",  // An exact version is required.  We do not permit pinning to a major or minor version.
            "detect": "a --version && cat /etc/a/.markerfile", // Only continue installation of this Feature if detect returns non-zero
            "options": {
                "bar": true
            }
        },
        {
            "id": "ghcr.io/microsoft/features/b",
            "version": "sha256:45b23dee08af5e43a7fea6c4cf9c25ccf269ee113168c19722f87876677c5cb2", // SHA of the published artifact is OK
            "detect": undefined, // Omit or set as 'undefined' to always install this Feature.
            "options": {
                "zip": "zap"
            }
        }
    ]
}
```

> An optional `finalize.sh` script can be included, and will be fired after all Features have been installed.

Reviewer 2: Do we have good examples for that?

Author: This could be omitted in the first pass if we find it to be unnecessary or without justification yet.  I'll leave for discussion

Reviewer 2: I like that we are discussing various options. We should aim for as much simplicity as we can while still covering the important cases we have. It will be easier to later add to a simple design than a complex one.

Reviewer 3: main

> A composite Feature will be published following the same process as an [standalone dev container Feature](https://containers.dev/implementors/features) into the same namespace - following the pattern outlined in [the Features distribution spec](https://containers.dev/implementors/features-distribution/). Dependencies of a composite Feature can be published to the same or different namespaces.

Reviewer: ```suggestion
A composite Feature will be published following the same process as a [standalone dev container Feature](https://containers.dev/implementors/features) into the same namespace - following the pattern outlined in [the Features distribution spec](https://containers.dev/implementors/features-distribution/). Dependencies of a composite Feature can be published to the same or different namespaces.
```

Reviewer: This section reminds me that we may want to update the Feature distribution docs with some note that distributing composite Features follows the same process as standalone Features.

> An example repository structure for a repo with one composite Feature and an standalone Feature can be found below:

Reviewer: ```suggestion
An example repository structure for a repo with one composite Feature and a standalone Feature can be found below:
```

```
$ tree 

├── src
│   ├── composite
│   │   ├── README.md
│   │   ├── devcontainer-feature.composite.json
│   │   ├── finalize.sh
│   ├── a
│   │   ├── README.md
│   │   ├── devcontainer-feature.json
│   │   └── install.sh
...
```

### Notes:

> - Composite Features cannot depend on other composite Features. This is to prevent a circular dependencies and deep dependency chains.

Reviewer: ```suggestion
- Composite Features cannot depend on other composite Features. This is to prevent circular dependencies and deep dependency chains.
```

> - The `dependsOn` array is in the order that the Features should be installed.  This is important for Features that depend on the existence of a file or directory created by a previous Feature.

Reviewer: Where does `dependsOn` come from (as this is the first time it's mentioned in this doc)? Is it up to the user to determine the order?

Author: This was from an earlier draft. Removing for now.

- The `detect` property is optional.  If omitted, the Feature will always be installed.
- The `options` property is optional.  If omitted, the default options will be passed to the Feature's `install.sh` script, as defined in the Feature's `devcontainer-feature.json` file.
- A composite Feature can optionally include a `finalize.sh` script.  This script will be executed after all of the dependencies have been installed.  This is useful for Features that need to perform some action after all of the dependencies have been installed.

## Advantages

- Composite Features can be used to distribute a single Feature that depends on other Features.
- Composite Features can pin all of their dependencies to a specific version, ensuring that the Feature can be tested and will work as expected.
- Composite Features prevent the complexity that arises with deeply nested dependencies or circular dependencies.

## Disadvantages

> - Composite features are not as flexible as other possible dependency models.

Reviewer: Do you have an example in mind of how other models are more flexible, so that readers can more fully understand which limitation(s) this refers to?
